rfcs/0182-taskcluster-yml-remote-references.md

```diff
@@ -0,0 +1,49 @@
+# RFC 182 - Allow remote references to .taskcluster.yml files processed by Taskcluster-GitHub
+* Comments: [#182](https://github.com/taskcluster/taskcluster-rfcs/pull/182)
+* Proposed by: @bhearsum
+
+# Summary
+
+Allow `.taskcluster.yml` files in GitHub repositories to outsource most of their contents to a `.taskcluster.yml` stored elsewhere.
+
+## Motivation
+
+`.taskcluster.yml` files in GitHub repositories are responsible for creating tasks (or not) in response to various GitHub events. In some cases (eg: when a decision task is used to delegate most of this to a separate tool), these files end up looking almost entirely identical, and hundreds of lines of JSON-e ends up duplicated across many different repositories. This duplication is a significant maintenance burden, and easily results in unwanted differences.
+
+Allowing a `.taskcluster.yml` to reference a file stored elsewhere will allow for deduplication of the bulk of the content in these files, resolving the concerns noted above.
+
+# Details
+
+Two new variables will be introduced to `.taskcluster.yml` files. The first, `config-from`, will allow an URL to be provided that points at a full-fledged `.taskcluster.yml` hosted elsewhere. The second, `context`, will allow for variables to be defined that will be included with the JSON-e context when the referenced `.taskcluster.yml` is rendered. Here is an example of a possible concrete `.taskcluster.yml`:
+
+```
+---
+version: 1
+config-from: https://github.com/taskcluster/taskgraph/blob/main/data/taskcluster-yml-github.yml
+context:
+  project-name: mozillavpn
+  scopes:
+    - secrets:get:project/mozillavpn/*
+```
+
+When `config-from` is present in the `.taskcluster.yml`, existing top level keys such as `policy` and `tasks` are not valid. `context` is optional, and only permitted when `config-from` is present.
```

Comment:
So does that mean that we wouldn't be able to have a mix of tasks from the "imported config" and tasks defined for the repository itself?

Reply:
This RFC doesn't allow for a formal way to do it -- but I think it would be _possible_ to do by specifying full task definitions in one of the `context` variables, if the referenced `.taskcluster.yml` allows for `extraTasks` or some such. (These extra tasks would not be able to use context of their own though, because that would require rendering the JSON-e twice...)

Something like this in the project `.taskcluster.yml`:
```
version: 1
config-from: https://....
context:
  extraTasks:
    - description: ...
      payload: ...
      ...
```

And this in the referenced one:
```
tasks:
  $flatten:
    - description: ...
      payload: ...
    - $eval: extraTasks
```

Reply:
I could certainly imagine this use case, yeah! This feels similar to the way that GitHub actions are sometimes composed (except in this case, we're merging a set of full tasks rather than allowing for individual steps to be remotely referenced?).

In the decision task case, I think this is mostly moot -- the sharing problem ends up being punted to whatever the decision task is (in our case, Taskgraph), and we have other ideas about how to solve it there.

In any case, if you feel it's important that this is included, I'm happy to update the RFC for it. It could also plausibly be an improvement later?

Reply:
I am not fully sure to be honest how to merge those when we would have multiple files. If they all produce `tasks[]` then we can probably just stack them and execute in any order. But if there would be some "meta-yamls" that define some variables that should be used by the rest, then it could get tricky. 
Maybe we should start with the most simple scenario, just one file for now

Reply:
It sounds to me like the tl;dr of this thread is that the current proposal is OK for now, but we could consider enhancements later. Given that, I'm going to resolve this thread -- please feel free to re-open (or tell me to) if any of that is wrong.